### ISSUE TITLE
installation problem in python3

### ISSUE THREAD

**Issue by dtenenba** [sirfz/tesserocr#21]
Hi, I'm trying to install tesserocr in a Docker container that's set up as follows:

```
FROM ubuntu:14.04

RUN apt-get -y update && apt-get install -y tesseract-ocr python3-imaging python3-pip python3-skimage libtesseract-dev libleptonica-dev

RUN pip3 install pytesseract ipython Cython
```

Then inside the container I manually run the command:

```
pip3 install tesserocr
```

It fails with the following. Any tips for how to get past this? Thanks.

```
Downloading/unpacking tesserocr
  Downloading tesserocr-2.1.3.tar.gz (49kB): 49kB downloaded
  Running setup.py (path:/tmp/pip_build_root/tesserocr/setup.py) egg_info for package tesserocr
    /usr/local/lib/python3.4/dist-packages/Cython/Distutils/old_build_ext.py:30: UserWarning: Cython.Distutils.old_build_ext does not properly handle dependencies and is deprecated.
      "Cython.Distutils.old_build_ext does not properly handle dependencies "
    Supporting tesseract v3.03
    Building with configs: {'libraries': ['tesseract', 'lept'], 'cython_compile_time_env': {'TESSERACT_VERSION': 771}}
Installing collected packages: tesserocr
  Running setup.py install for tesserocr
    /usr/local/lib/python3.4/dist-packages/Cython/Distutils/old_build_ext.py:30: UserWarning: Cython.Distutils.old_build_ext does not properly handle dependencies and is deprecated.
      "Cython.Distutils.old_build_ext does not properly handle dependencies "
    Supporting tesseract v3.03
    Building with configs: {'cython_compile_time_env': {'TESSERACT_VERSION': 771}, 'libraries': ['tesseract', 'lept']}
    cythoning tesserocr.pyx to tesserocr.cpp
    building 'tesserocr' extension
    x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/python3.4m -c tesserocr.cpp -o build/temp.linux-x86_64-3.4/tesserocr.o
    cc1plus: warning: command line option '-Wstrict-prototypes' is valid for C/ObjC but not for C++ [enabled by default]
    tesserocr.cpp: In function 'PyObject* __pyx_pf_9tesserocr_14PyPageIterator_20SetBoundingBoxComponents(__pyx_obj_9tesserocr_PyPageIterator*, bool, bool)':
    tesserocr.cpp:4610:25: error: 'class tesseract::PageIterator' has no member named 'SetBoundingBoxComponents'
       __pyx_v_self->_piter->SetBoundingBoxComponents(__pyx_v_include_upper_dots, __pyx_v_include_lower_dots);
                             ^
    tesserocr.cpp: In function 'PyObject* __pyx_pf_9tesserocr_14PyPageIterator_34GetImage(__pyx_obj_9tesserocr_PyPageIterator*, tesseract::PageIteratorLevel, int, PyObject*)':
    tesserocr.cpp:5842:125: error: no matching function for call to 'tesseract::PageIterator::GetImage(tesseract::PageIteratorLevel&, int&, Pix*&, int*, int*)'
       __pyx_v_pix = __pyx_v_self->_piter->GetImage(__pyx_v_level, __pyx_v_padding, __pyx_v_opix, (&__pyx_v_left), (&__pyx_v_top));
                                                                                                                                 ^
    tesserocr.cpp:5842:125: note: candidate is:
    In file included from tesserocr.cpp:424:0:
    /usr/include/tesseract/pageiterator.h:239:8: note: Pix* tesseract::PageIterator::GetImage(tesseract::PageIteratorLevel, int, int*, int*) const
       Pix* GetImage(PageIteratorLevel level, int padding,
            ^
    /usr/include/tesseract/pageiterator.h:239:8: note:   candidate expects 4 arguments, 5 provided
    tesserocr.cpp: In function 'PyObject* __pyx_pf_9tesserocr_13PyTessBaseAPI_74AnalyseLayout(__pyx_obj_9tesserocr_PyTessBaseAPI*, bool)':
    tesserocr.cpp:16244:83: error: no matching function for call to 'tesseract::TessBaseAPI::AnalyseLayout(bool&)'
       __pyx_v_piter = __pyx_v_self->_baseapi.AnalyseLayout(__pyx_v_merge_similar_words);
                                                                                       ^
    tesserocr.cpp:16244:83: note: candidate is:
    In file included from tesserocr.cpp:429:0:
    /usr/include/tesseract/baseapi.h:489:17: note: tesseract::PageIterator* tesseract::TessBaseAPI::AnalyseLayout()
       PageIterator* AnalyseLayout();
                     ^
    /usr/include/tesseract/baseapi.h:489:17: note:   candidate expects 0 arguments, 1 provided
    tesserocr.cpp: In function 'tesseract::TessResultRenderer* __pyx_f_9tesserocr_13PyTessBaseAPI__get_renderer(__pyx_obj_9tesserocr_PyTessBaseAPI*, __pyx_t_9tesseract_cchar_t*)':
    tesserocr.cpp:16588:88: error: no matching function for call to 'tesseract::TessHOcrRenderer::TessHOcrRenderer(__pyx_t_9tesseract_cchar_t*&, bool&)'
           __pyx_t_2 = new tesseract::TessHOcrRenderer(__pyx_v_outputbase, __pyx_v_font_info);
                                                                                            ^
    tesserocr.cpp:16588:88: note: candidates are:
    In file included from tesserocr.cpp:427:0:
    /usr/include/tesseract/renderer.h:175:3: note: tesseract::TessHOcrRenderer::TessHOcrRenderer()
       TessHOcrRenderer();
       ^
    /usr/include/tesseract/renderer.h:175:3: note:   candidate expects 0 arguments, 2 provided
    /usr/include/tesseract/renderer.h:173:16: note: tesseract::TessHOcrRenderer::TessHOcrRenderer(const tesseract::TessHOcrRenderer&)
     class TESS_API TessHOcrRenderer : public TessResultRenderer {
                    ^
    /usr/include/tesseract/renderer.h:173:16: note:   candidate expects 1 argument, 2 provided
    tesserocr.cpp:16631:106: error: no matching function for call to 'tesseract::TessPDFRenderer::TessPDFRenderer(__pyx_t_9tesseract_cchar_t*&, const char*)'
           __pyx_t_3 = new tesseract::TessPDFRenderer(__pyx_v_outputbase, __pyx_v_self->_baseapi.GetDatapath());
                                                                                                              ^
    tesserocr.cpp:16631:106: note: candidates are:
    In file included from tesserocr.cpp:427:0:
    /usr/include/tesseract/renderer.h:188:3: note: tesseract::TessPDFRenderer::TessPDFRenderer(const char*)
       TessPDFRenderer(const char *datadir);
       ^
    /usr/include/tesseract/renderer.h:188:3: note:   candidate expects 1 argument, 2 provided
    /usr/include/tesseract/renderer.h:186:16: note: tesseract::TessPDFRenderer::TessPDFRenderer(const tesseract::TessPDFRenderer&)
     class TESS_API TessPDFRenderer : public TessResultRenderer {
                    ^
    /usr/include/tesseract/renderer.h:186:16: note:   candidate expects 1 argument, 2 provided
    tesserocr.cpp:16715:69: error: no matching function for call to 'tesseract::TessUnlvRenderer::TessUnlvRenderer(__pyx_t_9tesseract_cchar_t*&)'
           __pyx_t_4 = new tesseract::TessUnlvRenderer(__pyx_v_outputbase);
                                                                         ^
    tesserocr.cpp:16715:69: note: candidates are:
    In file included from tesserocr.cpp:427:0:
    /usr/include/tesseract/renderer.h:227:3: note: tesseract::TessUnlvRenderer::TessUnlvRenderer()
       TessUnlvRenderer();
       ^
    /usr/include/tesseract/renderer.h:227:3: note:   candidate expects 0 arguments, 1 provided
    /usr/include/tesseract/renderer.h:225:16: note: tesseract::TessUnlvRenderer::TessUnlvRenderer(const tesseract::TessUnlvRenderer&)
     class TESS_API TessUnlvRenderer : public TessResultRenderer {
                    ^
    /usr/include/tesseract/renderer.h:225:16: note:   no known conversion for argument 1 from '__pyx_t_9tesseract_cchar_t* {aka const char*}' to 'const tesseract::TessUnlvRenderer&'
    tesserocr.cpp:16799:72: error: no matching function for call to 'tesseract::TessBoxTextRenderer::TessBoxTextRenderer(__pyx_t_9tesseract_cchar_t*&)'
           __pyx_t_5 = new tesseract::TessBoxTextRenderer(__pyx_v_outputbase);
                                                                            ^
    tesserocr.cpp:16799:72: note: candidates are:
    In file included from tesserocr.cpp:427:0:
    /usr/include/tesseract/renderer.h:238:3: note: tesseract::TessBoxTextRenderer::TessBoxTextRenderer()
       TessBoxTextRenderer();
       ^
    /usr/include/tesseract/renderer.h:238:3: note:   candidate expects 0 arguments, 1 provided
    /usr/include/tesseract/renderer.h:236:16: note: tesseract::TessBoxTextRenderer::TessBoxTextRenderer(const tesseract::TessBoxTextRenderer&)
     class TESS_API TessBoxTextRenderer : public TessResultRenderer {
                    ^
    /usr/include/tesseract/renderer.h:236:16: note:   no known conversion for argument 1 from '__pyx_t_9tesseract_cchar_t* {aka const char*}' to 'const tesseract::TessBoxTextRenderer&'
    tesserocr.cpp:16883:69: error: no matching function for call to 'tesseract::TessTextRenderer::TessTextRenderer(__pyx_t_9tesseract_cchar_t*&)'
           __pyx_t_6 = new tesseract::TessTextRenderer(__pyx_v_outputbase);
                                                                         ^
    tesserocr.cpp:16883:69: note: candidates are:
    In file included from tesserocr.cpp:427:0:
    /usr/include/tesseract/renderer.h:164:3: note: tesseract::TessTextRenderer::TessTextRenderer()
       TessTextRenderer();
       ^
    /usr/include/tesseract/renderer.h:164:3: note:   candidate expects 0 arguments, 1 provided
    /usr/include/tesseract/renderer.h:162:16: note: tesseract::TessTextRenderer::TessTextRenderer(const tesseract::TessTextRenderer&)
     class TESS_API TessTextRenderer : public TessResultRenderer {
                    ^
    /usr/include/tesseract/renderer.h:162:16: note:   no known conversion for argument 1 from '__pyx_t_9tesseract_cchar_t* {aka const char*}' to 'const tesseract::TessTextRenderer&'
    tesserocr.cpp: In function 'PyObject* __pyx_pf_9tesserocr_13PyTessBaseAPI_108IsValidCharacter(__pyx_obj_9tesserocr_PyTessBaseAPI*, PyObject*)':
    tesserocr.cpp:19649:60: error: 'class tesseract::TessBaseAPI' has no member named 'IsValidCharacter'
       __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_v_self->_baseapi.IsValidCharacter(__pyx_t_2)); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 2161, __pyx_L1_error)
                                                                ^
    tesserocr.cpp:532:36: note: in definition of macro '__Pyx_PyBool_FromLong'
     #define __Pyx_PyBool_FromLong(b) ((b) ? __Pyx_NewRef(Py_True) : __Pyx_NewRef(Py_False))
                                        ^
    tesserocr.cpp: In function 'PyObject* PyInit_tesserocr()':
    tesserocr.cpp:25002:67: error: 'PSM_RAW_LINE' is not a member of 'tesseract'
       __pyx_t_2 = __Pyx_PyInt_From_enum__tesseract_3a__3a_PageSegMode(tesseract::PSM_RAW_LINE); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 132, __pyx_L1_error)
                                                                       ^
    error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
    Complete output from command /usr/bin/python3 -c "import setuptools, tokenize;__file__='/tmp/pip_build_root/tesserocr/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-cwirkbsc-record/install-record.txt --single-version-externally-managed --compile:
    running install

running build

running build_ext

/usr/local/lib/python3.4/dist-packages/Cython/Distutils/old_build_ext.py:30: UserWarning: Cython.Distutils.old_build_ext does not properly handle dependencies and is deprecated.

  "Cython.Distutils.old_build_ext does not properly handle dependencies "

Supporting tesseract v3.03

Building with configs: {'cython_compile_time_env': {'TESSERACT_VERSION': 771}, 'libraries': ['tesseract', 'lept']}

cythoning tesserocr.pyx to tesserocr.cpp

building 'tesserocr' extension

creating build

creating build/temp.linux-x86_64-3.4

x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/python3.4m -c tesserocr.cpp -o build/temp.linux-x86_64-3.4/tesserocr.o

cc1plus: warning: command line option '-Wstrict-prototypes' is valid for C/ObjC but not for C++ [enabled by default]

tesserocr.cpp: In function 'PyObject* __pyx_pf_9tesserocr_14PyPageIterator_20SetBoundingBoxComponents(__pyx_obj_9tesserocr_PyPageIterator*, bool, bool)':

tesserocr.cpp:4610:25: error: 'class tesseract::PageIterator' has no member named 'SetBoundingBoxComponents'

   __pyx_v_self->_piter->SetBoundingBoxComponents(__pyx_v_include_upper_dots, __pyx_v_include_lower_dots);

                         ^

tesserocr.cpp: In function 'PyObject* __pyx_pf_9tesserocr_14PyPageIterator_34GetImage(__pyx_obj_9tesserocr_PyPageIterator*, tesseract::PageIteratorLevel, int, PyObject*)':

tesserocr.cpp:5842:125: error: no matching function for call to 'tesseract::PageIterator::GetImage(tesseract::PageIteratorLevel&, int&, Pix*&, int*, int*)'

   __pyx_v_pix = __pyx_v_self->_piter->GetImage(__pyx_v_level, __pyx_v_padding, __pyx_v_opix, (&__pyx_v_left), (&__pyx_v_top));

                                                                                                                             ^

tesserocr.cpp:5842:125: note: candidate is:

In file included from tesserocr.cpp:424:0:

/usr/include/tesseract/pageiterator.h:239:8: note: Pix* tesseract::PageIterator::GetImage(tesseract::PageIteratorLevel, int, int*, int*) const

   Pix* GetImage(PageIteratorLevel level, int padding,

        ^

/usr/include/tesseract/pageiterator.h:239:8: note:   candidate expects 4 arguments, 5 provided

tesserocr.cpp: In function 'PyObject* __pyx_pf_9tesserocr_13PyTessBaseAPI_74AnalyseLayout(__pyx_obj_9tesserocr_PyTessBaseAPI*, bool)':

tesserocr.cpp:16244:83: error: no matching function for call to 'tesseract::TessBaseAPI::AnalyseLayout(bool&)'

   __pyx_v_piter = __pyx_v_self->_baseapi.AnalyseLayout(__pyx_v_merge_similar_words);

                                                                                   ^

tesserocr.cpp:16244:83: note: candidate is:

In file included from tesserocr.cpp:429:0:

/usr/include/tesseract/baseapi.h:489:17: note: tesseract::PageIterator* tesseract::TessBaseAPI::AnalyseLayout()

   PageIterator* AnalyseLayout();

                 ^

/usr/include/tesseract/baseapi.h:489:17: note:   candidate expects 0 arguments, 1 provided

tesserocr.cpp: In function 'tesseract::TessResultRenderer* __pyx_f_9tesserocr_13PyTessBaseAPI__get_renderer(__pyx_obj_9tesserocr_PyTessBaseAPI*, __pyx_t_9tesseract_cchar_t*)':

tesserocr.cpp:16588:88: error: no matching function for call to 'tesseract::TessHOcrRenderer::TessHOcrRenderer(__pyx_t_9tesseract_cchar_t*&, bool&)'

       __pyx_t_2 = new tesseract::TessHOcrRenderer(__pyx_v_outputbase, __pyx_v_font_info);

                                                                                        ^

tesserocr.cpp:16588:88: note: candidates are:

In file included from tesserocr.cpp:427:0:

/usr/include/tesseract/renderer.h:175:3: note: tesseract::TessHOcrRenderer::TessHOcrRenderer()

   TessHOcrRenderer();

   ^

/usr/include/tesseract/renderer.h:175:3: note:   candidate expects 0 arguments, 2 provided

/usr/include/tesseract/renderer.h:173:16: note: tesseract::TessHOcrRenderer::TessHOcrRenderer(const tesseract::TessHOcrRenderer&)

 class TESS_API TessHOcrRenderer : public TessResultRenderer {

                ^

/usr/include/tesseract/renderer.h:173:16: note:   candidate expects 1 argument, 2 provided

tesserocr.cpp:16631:106: error: no matching function for call to 'tesseract::TessPDFRenderer::TessPDFRenderer(__pyx_t_9tesseract_cchar_t*&, const char*)'

       __pyx_t_3 = new tesseract::TessPDFRenderer(__pyx_v_outputbase, __pyx_v_self->_baseapi.GetDatapath());

                                                                                                          ^

tesserocr.cpp:16631:106: note: candidates are:

In file included from tesserocr.cpp:427:0:

/usr/include/tesseract/renderer.h:188:3: note: tesseract::TessPDFRenderer::TessPDFRenderer(const char*)

   TessPDFRenderer(const char *datadir);

   ^

/usr/include/tesseract/renderer.h:188:3: note:   candidate expects 1 argument, 2 provided

/usr/include/tesseract/renderer.h:186:16: note: tesseract::TessPDFRenderer::TessPDFRenderer(const tesseract::TessPDFRenderer&)

 class TESS_API TessPDFRenderer : public TessResultRenderer {

                ^

/usr/include/tesseract/renderer.h:186:16: note:   candidate expects 1 argument, 2 provided

tesserocr.cpp:16715:69: error: no matching function for call to 'tesseract::TessUnlvRenderer::TessUnlvRenderer(__pyx_t_9tesseract_cchar_t*&)'

       __pyx_t_4 = new tesseract::TessUnlvRenderer(__pyx_v_outputbase);

                                                                     ^

tesserocr.cpp:16715:69: note: candidates are:

In file included from tesserocr.cpp:427:0:

/usr/include/tesseract/renderer.h:227:3: note: tesseract::TessUnlvRenderer::TessUnlvRenderer()

   TessUnlvRenderer();

   ^

/usr/include/tesseract/renderer.h:227:3: note:   candidate expects 0 arguments, 1 provided

/usr/include/tesseract/renderer.h:225:16: note: tesseract::TessUnlvRenderer::TessUnlvRenderer(const tesseract::TessUnlvRenderer&)

 class TESS_API TessUnlvRenderer : public TessResultRenderer {

                ^

/usr/include/tesseract/renderer.h:225:16: note:   no known conversion for argument 1 from '__pyx_t_9tesseract_cchar_t* {aka const char*}' to 'const tesseract::TessUnlvRenderer&'

tesserocr.cpp:16799:72: error: no matching function for call to 'tesseract::TessBoxTextRenderer::TessBoxTextRenderer(__pyx_t_9tesseract_cchar_t*&)'

       __pyx_t_5 = new tesseract::TessBoxTextRenderer(__pyx_v_outputbase);

                                                                        ^

tesserocr.cpp:16799:72: note: candidates are:

In file included from tesserocr.cpp:427:0:

/usr/include/tesseract/renderer.h:238:3: note: tesseract::TessBoxTextRenderer::TessBoxTextRenderer()

   TessBoxTextRenderer();

   ^

/usr/include/tesseract/renderer.h:238:3: note:   candidate expects 0 arguments, 1 provided

/usr/include/tesseract/renderer.h:236:16: note: tesseract::TessBoxTextRenderer::TessBoxTextRenderer(const tesseract::TessBoxTextRenderer&)

 class TESS_API TessBoxTextRenderer : public TessResultRenderer {

                ^

/usr/include/tesseract/renderer.h:236:16: note:   no known conversion for argument 1 from '__pyx_t_9tesseract_cchar_t* {aka const char*}' to 'const tesseract::TessBoxTextRenderer&'

tesserocr.cpp:16883:69: error: no matching function for call to 'tesseract::TessTextRenderer::TessTextRenderer(__pyx_t_9tesseract_cchar_t*&)'

       __pyx_t_6 = new tesseract::TessTextRenderer(__pyx_v_outputbase);

                                                                     ^

tesserocr.cpp:16883:69: note: candidates are:

In file included from tesserocr.cpp:427:0:

/usr/include/tesseract/renderer.h:164:3: note: tesseract::TessTextRenderer::TessTextRenderer()

   TessTextRenderer();

   ^

/usr/include/tesseract/renderer.h:164:3: note:   candidate expects 0 arguments, 1 provided

/usr/include/tesseract/renderer.h:162:16: note: tesseract::TessTextRenderer::TessTextRenderer(const tesseract::TessTextRenderer&)

 class TESS_API TessTextRenderer : public TessResultRenderer {

                ^

/usr/include/tesseract/renderer.h:162:16: note:   no known conversion for argument 1 from '__pyx_t_9tesseract_cchar_t* {aka const char*}' to 'const tesseract::TessTextRenderer&'

tesserocr.cpp: In function 'PyObject* __pyx_pf_9tesserocr_13PyTessBaseAPI_108IsValidCharacter(__pyx_obj_9tesserocr_PyTessBaseAPI*, PyObject*)':

tesserocr.cpp:19649:60: error: 'class tesseract::TessBaseAPI' has no member named 'IsValidCharacter'

   __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_v_self->_baseapi.IsValidCharacter(__pyx_t_2)); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 2161, __pyx_L1_error)

                                                            ^

tesserocr.cpp:532:36: note: in definition of macro '__Pyx_PyBool_FromLong'

 #define __Pyx_PyBool_FromLong(b) ((b) ? __Pyx_NewRef(Py_True) : __Pyx_NewRef(Py_False))

                                    ^

tesserocr.cpp: In function 'PyObject* PyInit_tesserocr()':

tesserocr.cpp:25002:67: error: 'PSM_RAW_LINE' is not a member of 'tesseract'

   __pyx_t_2 = __Pyx_PyInt_From_enum__tesseract_3a__3a_PageSegMode(tesseract::PSM_RAW_LINE); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 132, __pyx_L1_error)

                                                                   ^

error: command 'x86_64-linux-gnu-gcc' failed with exit status 1

----------------------------------------
Cleaning up...
Command /usr/bin/python3 -c "import setuptools, tokenize;__file__='/tmp/pip_build_root/tesserocr/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-cwirkbsc-record/install-record.txt --single-version-externally-managed --compile failed with error code 1 in /tmp/pip_build_root/tesserocr
Storing debug log for failure in /root/.pip/pip.log
```

**Comment by sirfz** [comment]
tesserocr supports Tesseract v3.04 or greater while the version installed in your container is v3.03, you either need to manually compile an upgraded version or base your container on an upgraded distro which installs a more recent version of Tesseract.

**Comment by dtenenba** [comment]
Thanks. I was using Ubuntu 14.04 because another package I needed was not easily available for 16.04. However, excluding all other packages for the moment, I was able to get tesserocr running on Ubuntu 16.04 with the following `Dockerfile`:

```
FROM ubuntu:16.04

RUN apt-get -y update && apt-get install -y tesseract-ocr libtesseract-dev libleptonica-dev python3-pip

RUN pip3 install pillow ipython Cython

RUN pip3 install tesserocr

```

So I am sorted out for now. Thanks again.
